Allow configuring ASO for non-public cloud at install

nojnhuh opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind bug

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:

This PR allows for configuring CAPZ's ASO installation to point to a non-public cloud (like GovCloud) at CAPZ-install time using the ASO environment variables defined here: https://azure.github.io/azure-service-operator/guide/aso-controller-settings-options/.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [X] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
ASO environment variables can now be set to refer to a non-public (e.g. government) cloud instance: https://azure.github.io/azure-service-operator/guide/aso-controller-settings-options/
```

nojnhuh commented on config/aso/credentials.yaml:
```diff
@@ -3,7 +3,10 @@ kind: Secret
 type: Opaque
-data:
+stringData:
+  AZURE_AUTHORITY_HOST: ${AZURE_AUTHORITY_HOST:=""}
```
Where should we document these? I don't see any existing docs about how to define this kind of config today.

ionutleca replied:
I've been using CAPZ to target _management.usgovcloudapi.net_  using a local  infrastructure-components.yaml with those same 3 secret values set.

I think you already tested what needed to be tested for this PR if `clusterctl init` resulted in a pacthed aso secret ✔️.

I will be able to delete some code on my side if this goes in, thanks 😄

ionutleca replied:
Not sure where in this repo, but this might be a good place to also mention them https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/user/quick-start.md?plain=1#L455

nojnhuh replied:
@dtzar Did you know of a good spot to document these settings? Or could we defer this for now here and let that be part of a larger docs refactor?

nojnhuh commented:
@ionutleca If you're able to test these changes I'd appreciate it. I can get as far as defining these and hitting an error, but my sub doesn't have quote for other clouds where I could test that this is actually working.

ionutleca commented:
@nojnhuh, I've been using CAPZ to target _management.usgovcloudapi.net_  using a local  infrastructure-components.yaml with those same 3 secret values set.

I think you already tested what needed to be tested for this PR if `clusterctl init` resulted in a pacthed aso secret ✔️.

I will be able to delete some code on my side if this goes in, thanks 😄

nojnhuh commented:
@CecileRobertMichon Is this fair to cherry-pick?

/cherry-pick release-1.11
/cherry-pick release-1.12

willie-yao reviewed (commented):
Looks good to me pending some documentation for the variables.

dtzar commented:
I would recommend having a new section of something like **Configuration** in the [ASO topic book page](https://capz.sigs.k8s.io/topics/aso).

dtzar commented:
Can we also please add the other relevant ASO configuration options?  At a minimum, it seems we should add:

```
USE_WORKLOAD_IDENTITY_AUTH
AZURE_SYNC_PERIOD
AZURE_USER_AGENT_SUFFIX
```
Without the first one, ASO will not be able to utilize workload identity auth.

These ones are all tied to multi-tenancy and since today CAPZ doesn't support this mode, it does make sense to call out and not make it easy to configure them.
```
AZURE_OPERATOR_MODE
AZURE_TARGET_NAMESPACES 
```

nojnhuh commented:
> Can we also please add the other relevant ASO configuration options? At a minimum, it seems we should add:
> 
> ```
> USE_WORKLOAD_IDENTITY_AUTH
> AZURE_SYNC_PERIOD
> AZURE_USER_AGENT_SUFFIX
> ```
> 
> Without the first one, ASO will not be able to utilize workload identity auth.

Workload identity works between CAPZ and ASO today. CAPZ configures per-resource secrets for auth which include the necessary workload identity config. This option only applies to the default global authentication config for ASO which is superseded by the per-resource secrets that CAPZ creates. Are you seeing workload identity not working?

I'll also add `AZURE_SYNC_PERIOD` here. I'd vote to wait for ASO v2.6.0 to be released though before we add config for `AZURE_USER_AGENT_SUFFIX` since that's not published in a release yet.

dtzar commented:
ASO is provisioning resources fine with my CAPZ which has WI setup, but I haven't validated how it's doing this.

`Workload identity works between CAPZ and ASO today. CAPZ configures per-resource secrets for auth which include the necessary workload identity config. This option only applies to the default global authentication config for ASO which is superseded by the per-resource secrets that CAPZ creates.`  
This would be a good note to have - not necessarily in this PR - because the [ASO instructions to setup WI ](https://azure.github.io/azure-service-operator/guide/authentication/credential-format/#create-the-secret) shows you have to set this variable to true.

nojnhuh commented:
/hold for squash
/retitle Allow configuring ASO settings at CAPZ install

willie-yao reviewed (approved):
/lgtm

dtzar commented on docs/book/src/topics/configuration.md:
```diff
@@ -0,0 +1,26 @@
+# Configuration
+
+This page describes how to configure CAPZ installations.
```
This is a very broad statement at the root level documentation.  My suggestion was just under the ASO topic area.  At this level, I'd expect much more explanation with all the variables for CAPZ/ASO, what's required/default, and links or included to setup things like multi-tenancy, etc.

dtzar commented on docs/book/src/topics/configuration.md:
```diff
@@ -0,0 +1,26 @@
+
+## Environment Variables
+
+These environment variables are used to template the kustomize output for CAPZ control plane installations.
```
Would be good to let people know they must set these environment variables before `clusterctl init` OR if they need to modify/add them afterwards how to do so.

CecileRobertMichon commented:
/lgtm
/approve

nojnhuh commented:
/release-note-edit
```release-note
ASO environment variables can now be set to refer to a non-public (e.g. government) cloud instance and set ASO's periodic resync interval: https://azure.github.io/azure-service-operator/guide/aso-controller-settings-options/
```

nojnhuh commented:
Looking into why LGTM got dropped.

nojnhuh commented:
Fixed and squashed!
/hold cancel

@CecileRobertMichon PTAL, the tree hash matches again:
```
% git rev-parse HEAD:
fc4b435749bd0422c90775b816324717c3749ecb
```

CecileRobertMichon reviewed (commented):
/lgtm
